acme/acme/messages.py

```diff
@@ -258,6 +259,27 @@ class ResourceBody(jose.JSONObjectWithFields):
     """ACME Resource Body."""
 
 
+class ExternalAccountBinding:
+    """ACME External Account Binding"""
+
+    @classmethod
+    def from_data(cls, key=None, kid=None, hmac=None):
+        """Create External Account Binding Resource from contact details, kid and hmac."""
+
+        key_json = json.dumps(key.to_partial_json())
+        decoded_hmac = jose.b64.b64decode(hmac)
+        jws = jose.JWS.sign(
+            payload=key_json,
+            key=jose.jwk.JWKOct(key=decoded_hmac),
+            alg=jose.jwa.HS256,
+            kid=kid,
+            include_jwk=False,
+            protect=frozenset(['alg', 'kid'])
```

Comment:
I've now implemented the functionality that makes Certbot error out if the server requires external account binding and the user didn't specify the kid and hmac key on command line. It's in commit: https://github.com/certbot/certbot/pull/6059/commits/854804f87d10cb6d35a7b29c1942f2346cffd6f0

By raising errors.Error in the code it seems that the exception only gets logged in the logfile and not outputted to stdout. Instead the generic error message "Unable to register an account with ACME server". This is the code: https://github.com/certbot/certbot/blob/36ebce4a5fa30a0e73bd8656aea54ded2e9d982e/certbot/main.py#L527

Would this be in line with the Certbot logging philosophy? Or do we need to create a separate exception for this error to make the real error message show up on stdout?

Reply:
To translate: that assertion is saying that there should be a line in https://github.com/certbot/josepy/blob/b0ad174768af90efbf6c98e9d68fa201108e9386/src/josepy/jws.py#L38 that says something like `url = json_util.Field('url', omitempty=True)`

What is a mystery to me is how the other code you reference is working at all without that line... any thoughts?

Reply:
ah yes, there we go

`Header` is subclassed here: https://github.com/certbot/certbot/blob/master/acme/acme/jws.py#L10

instead of using `jose.JWS` directly, do what `client.py` does and `from acme import jws` and then use `jws.JWS`

Reply:
Thanks a lot! Will look into this now.